pykafka/rdkafka/producer.py

```diff
@@ -0,0 +1,133 @@
+import logging
+
+from pykafka.producer import Producer, CompressionType
+from pykafka.handlers import ResponseFuture
+from . import _rd_kafka
+
+
+log = logging.getLogger(__name__)
+
+
+class RdKafkaProducer(Producer):
+    """A librdkafka-backed version of pykafka.Producer
+
+    This aims to conform to the Producer interface as closely as possible.  One
+    notable difference is that it returns a future from produce(), through
+    which the user can later check if produced messages have made it to kafka.
+
+    For an overview of how configuration keys are mapped to librdkafka's, see
+    _mk_rdkafka_config_lists.
+    """
+    def start(self):
+        if not self._running:
+            brokers = ','.join(map(lambda b: ':'.join((b.host, str(b.port))),
+                                   self._cluster.brokers.values()))
+            conf, topic_conf = self._mk_rdkafka_config_lists()
+
+            self._rdk_producer = _rd_kafka.Producer()
+            self._rdk_producer.configure(conf=conf)
+            self._rdk_producer.configure(topic_conf=topic_conf)
+            self._rdk_producer.start(brokers, self._topic.name)
+            self._running = True
+
+            # TODO spawn thread to regularly poll() for delivery reports (and
+            # break if not _running)
+
+    def _produce(self, message_partition_tup):
+        # Delivery-reports will be available through this future
+        future = ResponseFuture(self._cluster.handler)
+
+        (key, msg), part_id = message_partition_tup
+        self._rdk_producer.produce(msg, key, part_id, future)
+        return future
```

Comment:
Thanks, yes that's precisely the reason. I was wondering about the same thing - would there be any better way?

I suspect that most users who'd want to evaluate the futures returned (to act on errors) would want to bulk-process them (and with `concurrent.futures` some facilities for that [come provided][0]) in a pattern roughly like so:
```python
# Warning: potential pseudo-code ahead
pending = []
while True:
    for _ in xrange(big_number):
        msg = rawdata_consumer.consume()
        result = do_stuff_with_msg(msg)
        future = nextstage_producer.produce(result)
        pending.append(future)
    timeout = 0  # ie don't wait, just grab what's been delivered
    done, not_done = concurrent.futures.wait(pending, timeout)
    errors = [f.exception() for f in done if f.exception() is not None]
    pending = not_done
```

... and so, one might argue that if that's going to be the pattern, you ought to abstract the pattern into the producer interface, to save everyone the hassle of writing it all the time. On the other hand, it's not a terribly big or complicated pattern, and not baking it into the interface is more flexible, and we can always add convenience-wrappers for this later? So... I guess I'm starting to lean the way of just returning a future (but not a `handlers.ResponseFuture`, which would lack all the stuff around it).

[0]: https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.wait